File: algorithms/python/createSortedArraythroughInstructions/createSortedArraythroughInstructions.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

# Source: https://leetcode.com/problems/create-sorted-array-through-instructions/
# Author: Miao Zhang
# Date:   2021-05-25

class FenwickTree:
    def __init__(self, n):
        self.sums = [0 for _ in range(n + 1)]
    
    def update(self, i, delta):
        while i < len(self.sums):
            self.sums[i] += delta
            i += self.lowbit(i)
    
    def query(self, i):
        res = 0
        while i > 0:
            res += self.sums[i]
            i -= self.lowbit(i)
        return res
    
    def lowbit(self, x):
        return x & (-x)
    

class Solution:
    def createSortedArray(self, instructions: List[int]) -> int:
        n = len(instructions)
        kMod = 10 ** 9 + 7
        tree = FenwickTree(10 ** 5)
        res = 0
        for i in range(n):
            lt = tree.query(instructions[i] - 1)
            gt = i - tree.query(instructions[i])
            res += min(lt, gt)
            tree.update(instructions[i], 1)
        return res % kMod

File: algorithms/cpp/createSortedArraythroughInstructions/createSortedArraythroughInstructions.cpp
// Source: https://leetcode.com/problems/create-sorted-array-through-instructions/
// Author: Miao Zhang
// Date:   2021-05-25

class FenwickTree {
public:
    FenwickTree(int n): sums_(n + 1) {}

    void update(int i, int delta) {
        while (i < sums_.size()) {
            sums_[i] += delta;
            i += lowbit(i);
        }
    }

    int query(int i) const {
        int sums = 0;
        while (i > 0) {
            sums += sums_[i];
            i -= lowbit(i);
        }
        return sums;
    }

private:
    static inline int lowbit(int x) {
        return x & (-x);
    }
    vector<int> sums_;
};


class Solution {
public:
    int createSortedArray(vector<int>& instructions) {
        int kMod = 1e9 + 7;
        int n = instructions.size();
        FenwickTree tree(1e5);
        long res = 0;
        for (int i = 0; i < n; i++) {
            int lt = tree.query(instructions[i] - 1);
            int gt = i - tree.query(instructions[i]);
            res += min(lt, gt);
            tree.update(instructions[i], 1);
        }
        return res % kMod;
    }
};
